Features/frontend default build

scheibelp opened the pull request:
A few notes:

-whether a dependency is a build dep is determined by checking whether one can reach the root by traversing link deps
-link dependencies of build deps should be built with the frontend compiler
-if a parent links to a dependency of a build dep however, that may not always be the case (I think this is mostly handled)

Also this appears to significantly boost concretization time as-is.

pramodk commented:
Quickly trying example discussed in #1980 : 

```bash
spack spec parmetis ^mpich
Input spec
--------------------------------
parmetis
    ^mpich

Normalized
--------------------------------
parmetis
    ^cmake@2.8:
    ^metis@5:
    ^mpich

Concretized
--------------------------------
parmetis@4.0.3%gcc@4.4.7~debug~gdb+shared arch=bgq-cnk-powerpc
    ^cmake@3.6.1%gcc@4.4.7~doc+ncurses+openssl~ownlibs~qt arch=bgq-rhel6-power7
        ^bzip2@1.0.6%gcc@4.4.7 arch=bgq-rhel6-power7
        ^curl@7.50.3%gcc@4.4.7 arch=bgq-rhel6-power7
            ^openssl@1.0.2j%gcc@4.4.7 arch=bgq-rhel6-power7
                ^zlib@1.2.8%gcc@4.4.7 arch=bgq-rhel6-power7
        ^expat@2.2.0%gcc@4.4.7 arch=bgq-rhel6-power7
        ^libarchive@3.2.1%gcc@4.4.7 arch=bgq-rhel6-power7
            ^libxml2@2.9.4%gcc@4.4.7~python arch=bgq-rhel6-power7
                ^xz@5.2.2%gcc@4.4.7 arch=bgq-rhel6-power7
            ^lz4@131%gcc@4.4.7 arch=bgq-rhel6-power7
            ^lzma@4.32.7%gcc@4.4.7 arch=bgq-rhel6-power7
            ^lzo@2.09%gcc@4.4.7 arch=bgq-rhel6-power7
            ^nettle@3.2%gcc@4.4.7 arch=bgq-rhel6-power7
                ^gmp@6.1.1%gcc@4.4.7 arch=bgq-rhel6-power7
                    ^m4@1.4.17%gcc@4.4.7+sigsegv arch=bgq-rhel6-power7
                        ^libsigsegv@2.10%gcc@4.4.7 arch=bgq-rhel6-power7
        ^ncurses@6.0%gcc@4.4.7 arch=bgq-rhel6-power7
    ^metis@5.1.0%gcc@4.4.7~debug~gdb~idx64~real64+shared arch=bgq-cnk-powerpc
    ^mpich@3.2%gcc@4.4.7+hydra+pmi+romio~verbs arch=bgq-cnk-powerpc
```
So this looks good! thank you @scheibelp!

scheibelp commented:
Thanks for checking that!

It looks like in this case this chose to compile everything with gcc@4.4.7

I may be able to tweak this so things that arent build deps do their own separate resolution. In the meantime, specifying xl for the top-level dep should get the result you are looking for.

scheibelp commented:
I just noticed that I'm not detecting whether a package is a build dep correctly 100% of the time. Once I fix that I expect this to work

scheibelp commented:
Looking into this, I appear to have confused myself and stumbled upon a bug (namely I was explicitly mentioning cmake in a spec and it showed up as a link deptype). I'll report the issue shortly. In the meantime I expect this to work as intended. I think it slows down concretization by 20-30%.

pramodk commented:
Looks like there is a bug now : 

```bash
./bin/spack spec parmetis
Input spec
--------------------------------
parmetis

Normalized
--------------------------------
parmetis
    ^cmake@2.8:
    ^metis@5:
    ^mpi

Concretized
--------------------------------
==> Error: Multiple compilers satisfy spec xl@12.1
```

This is when I add `xl` as preferred compiler in `~/.spack/bgq/packages.yaml` as:

```bash
packages:
    all:
        compiler: [xl,gcc]
```

It's true that there are `multiple` compilers satisfying spec `xl@12.1` but they are for different os (fronted-backend) : 

```bash
$ cat ~/.spack/bgq/compilers.yaml
compilers:
- compiler:
    flags: {}
    modules: []
    operating_system: rhel6
    paths:
      cc: /usr/lib64/ccache/gcc
      cxx: /usr/lib64/ccache/g++
      f77: /usr/bin/gfortran
      fc: /usr/bin/gfortran
    spec: gcc@4.4.7
- compiler:
    flags: {}
    modules: []
    operating_system: cnk
    paths:
      cc: /usr/lib64/ccache/gcc
      cxx: /usr/lib64/ccache/g++
      f77: /usr/bin/gfortran
      fc: /usr/bin/gfortran
    spec: gcc@4.4.7
- compiler:
    flags: {}
    modules: []
    operating_system: rhel6
    paths:
      cc: /opt/ibmcmp/vacpp/bg/12.1/bin/xlc_r
      cxx: /opt/ibmcmp/vacpp/bg/12.1/bin/xlc++
      f77: /opt/ibmcmp/xlf/bg/14.1/bin/xlf_r
      fc: /opt/ibmcmp/xlf/bg/14.1/bin/xlf2008
    spec: xl@12.1
- compiler:
    flags: {}
    modules: []
    operating_system: cnk
    paths:
      cc: /opt/ibmcmp/vacpp/bg/12.1/bin/xlc_r
      cxx: /opt/ibmcmp/vacpp/bg/12.1/bin/xlc++
      f77: /opt/ibmcmp/xlf/bg/14.1/bin/xlf_r
      fc: /opt/ibmcmp/xlf/bg/14.1/bin/xlf2008
    spec: xl@12.1
```
Note that this doesn't happen with `gcc` even though there are multiple `spec`s.

scheibelp commented:
Thanks for checking this I will try to look at this tonight.

pramodk commented:
@scheibelp : I should mention few additional details : 

* On our system I am seeing #2027. Hence I was previously use some additional local changes that I have in my branch.
* I am starting from beginning with below steps : 
 * Use latest  develop branch
 * Use following three PR's  to make spack working on `bg-q`
 ```bash
 git fetch upstream pull/2292/head:bgq_frontend
 git fetch upstream pull/1980/head:bgq_os
 git fetch upstream pull/2189/head:yaml_json

  merge all above three PRs with develop. Resolve conflict with `lib/spack/spack/spec.py` (from yaml related PR).
```
 * here is the [branch](https://github.com/pramodk/spack/tree/bgq_develop)  the branch that I created using above steps.

Now trying `spack ` : 

```bash

○ → rm -rf ~/.spack

○ → spack compiler find
spack arch==> Added 2 new compilers to /somepath/home/kumbhar/.spack/bgq/compilers.yaml
    gcc@4.4.7  gcc@4.4.7

 ○ → spack arch
bgq-cnk-ppc64

 ○ → spack spec parmetis
Input spec
--------------------------------
parmetis

Normalized
--------------------------------
parmetis
    ^cmake@2.8:
    ^metis@5:
    ^mpi

Concretized
--------------------------------
Traceback (most recent call last):
  File "/somepath//spack/bin/spack", line 195, in <module>
    main()
  File "/somepath//spack/bin/spack", line 172, in main
    return_val = command(parser, args)
  File "/somepath//spack/lib/spack/spack/cmd/spec.py", line 78, in spec
    spec.concretize()
  File "/somepath//spack/lib/spack/spack/spec.py", line 1354, in concretize
    self._concretize_helper())
  File "/somepath//spack/lib/spack/spack/spec.py", line 1175, in _concretize_helper
    changed |= dep.spec._concretize_helper(presets, visited)
  File "/somepath//spack/lib/spack/spack/spec.py", line 1185, in _concretize_helper
    (spack.concretizer.concretize_architecture(self),
  File "/somepath//spack/lib/spack/spack/concretize.py", line 309, in concretize_architecture
    self._concretize_operating_system(spec),
  File "/somepath//spack/lib/spack/spack/concretize.py", line 249, in _concretize_operating_system
    if spec.build_dep():
  File "/somepath//spack/lib/spack/spack/spec.py", line 608, in __getattr__
    raise AttributeError()
AttributeError
```
I don't have any other changes in `spack` install directory.  Am I missing anything here? Or should test somewhat differently?

pramodk commented:
Oops....wait a minute, there is something I have in my config: 

```bash

$ rm -rf ~/.spack
$ ./bin/spack config get compilers
compilers:
- compiler:
    modules: []
    operating_system: CNK
    paths:
      cc: /usr/bin/bgxlc_r
      cxx: /usr/bin/bgxlc++
      f77: /usr/bin/bgxlf_r
      fc: /usr/bin/bgxlf2008
    spec: xl@12.1
```

So certainly `CNK` is something that I have locally. Let me fix that!

scheibelp commented:
```
  File "/somepath//spack/lib/spack/spack/concretize.py", line 249, in _concretize_operating_system
    if spec.build_dep():
  File "/somepath//spack/lib/spack/spack/spec.py", line 608, in __getattr__
    raise AttributeError()
```

```build_dep``` is a function I added for this PR. When I check spec.py in the branch you linked (https://github.com/pramodk/spack/tree/bgq_develop) I don't see this function. Perhaps this was omitted during your merge?

Based on the above it seems like the issue could in fact have been caused by duplicate compiler definition. I'll revert the debugging commit for now and I can bring it back if the problem returns.

Thanks!

pramodk commented:
@scheibelp : I was too quick to merge different PR's and I made mistake! Sorry!
Everything looks fine with this PR. XL and GCC concretisation preferences are working fine! 

```bash

# with compiler: [xl, gcc]

$ ./bin/spack spec parmetis ^mpich
Input spec
--------------------------------
parmetis
    ^mpich

Normalized
--------------------------------
parmetis
    ^cmake@2.8:
    ^metis@5:
    ^mpich

Concretized
--------------------------------
parmetis@4.0.3%xl@12.1~debug~gdb+shared arch=bgq-cnk-ppc64
    ^cmake@3.6.1%xl@12.1~doc+ncurses+openssl~ownlibs~qt arch=bgq-rhel6-power7
        ^bzip2@1.0.6%xl@12.1 arch=bgq-rhel6-power7
        ^curl@7.50.3%xl@12.1 arch=bgq-rhel6-power7
            ^openssl@1.0.2j%xl@12.1 arch=bgq-rhel6-power7
                ^zlib@1.2.8%xl@12.1 arch=bgq-rhel6-power7
        ^expat@2.2.0%xl@12.1 arch=bgq-rhel6-power7
        ^libarchive@3.2.1%xl@12.1 arch=bgq-rhel6-power7
            ^libxml2@2.9.4%xl@12.1~python arch=bgq-rhel6-power7
                ^xz@5.2.2%xl@12.1 arch=bgq-rhel6-power7
            ^lz4@131%xl@12.1 arch=bgq-rhel6-power7
            ^lzma@4.32.7%xl@12.1 arch=bgq-rhel6-power7
            ^lzo@2.09%xl@12.1 arch=bgq-rhel6-power7
            ^nettle@3.2%xl@12.1 arch=bgq-rhel6-power7
                ^gmp@6.1.1%xl@12.1 arch=bgq-rhel6-power7
                    ^m4@1.4.17%xl@12.1+sigsegv arch=bgq-rhel6-power7
                        ^libsigsegv@2.10%xl@12.1 arch=bgq-rhel6-power7
        ^ncurses@6.0%xl@12.1 arch=bgq-rhel6-power7
    ^metis@5.1.0%xl@12.1~debug~gdb~idx64~real64+shared arch=bgq-cnk-ppc64
    ^mpich@3.2%xl@12.1+hydra+pmi+romio~verbs arch=bgq-cnk-ppc64
```

scheibelp commented:
Great! Thanks again for checking this! I'll work on doing clean up and add tests for this.

xjrc commented:
Just as a heads up, it looks like this PR and #2261 are going to have some non-trivial conflicts in `spack/spec/concretize.py`.  If #2261 ends up getting merged first, I'll follow up here to help out with resolving the conflicts.

scheibelp commented:
@xjrc Thanks for that! 

A cursory glance tells me it shouldn't be too bad: the more complex logic was in concretize_compiler and it doesnt appear that it will be affected. I'll just have to translate my logic to pick out the frontend os/target for build deps in concretize compiler (and I think I can see how that would look).

In any case I should be able to get this up to date within a few hours of #2261 merging.

scheibelp commented:
I think this is in a good spot now so I'll remove the WIP tag. I can update and resolve conflicts when #2261 is merged

tgamblin commented:
@scheibelp: #2261 is merged now. Can you rebase this one?

@becker33 See this PR for the `0.10` release.

scheibelp commented:
@becker33 IIRC you mentioned taking a crack at merging this yourself. Did you start on that? If not I can do that by tomorrow.

becker33 commented:
@scheibelp I think it may be easier for you to do it, since you know your code better than I will. I've started to familiarize myself with, but I'm not there yet..

scheibelp commented:
Makes sense! I'll do the rebase

scheibelp commented:
Rebase complete

I'll note from the commit message:

> This also treats run dependencies the same as build dependencies, so they are also built against the frontend by default and get compiler information from build/run deps.

I can alter that if desired

becker33 commented:
I believe we want run dependencies to default similarly to link dependencies.

scheibelp commented:
OK I'll update to that behavior (it will be ready this afternoon)

scheibelp commented:
Behavior is now updated to only prefer frontend for build deps (or dependencies of build deps that are not link/run dependencies for the root). Commit message also updated.

tgamblin commented:
@scheibelp: Thanks!

mpbelhorn commented:
@tgamblin Yes, this would probably confuse our Spack instances if we updated Spack in-place. But that's OK, because we don't update Spack in-place.

----

For anyone who has concerns about this sort of potential problem, the following is how we avoid it.

When we sync with upstream, we rebuild the entire software stack that we provide in a new location. Previously, this was an entirely different Spack instance in a new location, however the newly configurable `install_tree` should allow us to update Spack in-place, but 'abandon' previously installed packages in their own `install_tree`.

Leaving the older installed packages in-place on disk, but ignored by Spack allows us to avoid discontinuities regarding provided software and no worry about name clashes/changes. Our Spack deploys are hidden from users. We point them to installed packages through customized modulefiles in  a public place and only update them to point to the new stack once everything is built anew.

scheibelp commented:
> I just realized post-merging that this might affect the Cray folks.

This change can be reverted with no effect. I made it because while looking into something I noticed the compilers for CNL weren't found in the tests (the tests do not actually set the CNL as the backend anymore). I reverted this change at: https://github.com/LLNL/spack/pull/2526

tgamblin commented:
@mpbelhorn @mamelara: Thanks -- that's helpful.

Question: is it worth versioning the CNL OS?  Do you have CNL upgrades on Titan/other machines that would change the binary compatibility of the OS, and can we get the version in a deterministic way?  Currently the CNL OS is hard-coded at version `10` for the Cray machines.

mpbelhorn commented:
I'm aware that the version can change, but it doesn't change often in the lifetime of our machines. I'm not the one to ask - I don't even know what version of CNL is on our machines or  exactly where to find it.

Titan for example runs a modified SuSE11.3, ["Cray Linux Environment" CLE 5.2](http://docs.cray.com/cgi-bin/craydoc.cgi?mode=View;id=S-2425-52;right=/books/S-2425-52/html-S-2425-52//chapter-introduction.html%23section.2.hHhCoiBr) on the login/batch nodes. CLE is clearly versioned and releases ship with an associated (non-SuSE-based?)  "Compute Node Linux" CNL for the compute nodes but I don't know what the 'official' CNL version is.

I prefer to consider CLE simply SuSE11.
